
<!--==========================================================================
	FTL macros 
	========================================================================== -->


<!--==========================================================================
	CSS styles 
	========================================================================== -->


<style type="text/css">

body {
	background-color: #dddddd;
}

/* Table styles. */

table {
  border-color: #000;
  border-spacing: 0px;
  border-style: solid;
  border-width: 1px;
  cell-spacing: 0px;
}

.noBorder {
  border-width: 0px;
}

td, th {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 10pt;
  padding: 2px 0.5em;
  white-space: nowrap;
}

td.numeric {
  text-align: right;
}

th {
  background-color: #c0c0c0;
}

th.mainHeader {
  background-color: #808080;
  color: #ffffff;
  text-align: left;
}

th a {
  color: #000080;
  text-decoration: none;
}

th a:visited {
  color: #000080;
}

th a:active, th a:hover {
  color: #800000;
  text-decoration: underline;
}

.toc {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 95%;
}

.note {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 90%;
	text-align: left;
}

.main {
	margin-top: 15px;
	width: 1000px;
	text-align: left;
	background: #ffffff;
	border: 3px solid #777777;
}

.histo {
	table-layout:fixed;
	width:100%;
	border:1px solid #aaa;
	word-wrap:break-word;
}

</style>


<!--==========================================================================
	Index
	========================================================================== -->

<center>
<div class="main">

<center> <h3> SnpEff: Variant analysis </h3> </center>

<div style="margin-left: .5em">
<table class="toc"><tr><td>
	<center><b>Contents</b></center>
	<a href="#summary">Summary</a><br>
	<a href="#changeRateChr"> Change rate by chromosome</a><br>
	<a href="#changesByType">Variants by type</a><br>
	<a href="#effectsImpact"> Number of variants by impact </a><br>
	<a href="#effectsImpact"> Number of variants by functional class </a><br>
	<a href="#effects"> Number of variants by effect </a><br>
	<a href="#quality">Quality histogram</a><br>
	<a href="#coverage">Coverage histogram</a><br>
	<a href="#baseChages">Base change table</a><br>
	<a href="#tstv">Transition vs transversions (ts/tv)</a><br>
	<a href="#alleleFreq"> Frequency of alleles </a><br>
	<a href="#codonChanges"> Codon change table </a><br>
	<a href="#aaChanges"> Amino acid change table </a><br>
	<a href="#chrChanges"> Chromosome change plots </a><br>
	<a href="snpEff_genes.txt"> Details by gene </a><br>
</tr></td></table>
</div>


<!--==========================================================================
	Summary table 
	========================================================================== -->

<hr>
<a name="summary">
<center>
<b>Summary</b><p>

<table border=0>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Genome </b> </td>
		<td> maizeZmB73 </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Date </b> </td>
		<td> 2014-04-09 20:13 </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> SnpEff version </b> </td>
		<td> <pre>SnpEff 3.5 (build 2014-04-07), by Pablo Cingolani</pre> </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Command line arguments </b> </td>
		<td> <pre>SnpEff  maizeZmB73 /home/pcingola/snpEff/franz_output.vcf 
</pre> </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Warnings </b> </td>
		   
		<td bgcolor="#ff0000"> 59 </td>
	</tr>	
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of lines (input file) </b> </td>
		<td> 24 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of variants (before filter) </b> </td>
		<td> 37 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Filter </b> </td>
		<td>  </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants filtered out </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of not variants <br>(i.e. reference equals alternative) </b> </td>
		<td> 13 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants processed <br> (i.e. after filter and non-variants) </b> </td>
		<td> 24 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of known variants <br>(i.e. non-empty ID) </b> </td>
		<td> 
		1
		( 4.167% ) 
		</td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of effects </b> </td>
		<td> 125 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Genome total length </b> </td>
		<td> 2,066,432,719 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Genome effective length </b> </td>
		<td> 301,354,136 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Change rate </b> </td>
		<td> 1 change every 12,556,422 bases </td>
	</tr>
</table>
<p>
</center>

<!--==========================================================================
	Change rate by chromosome
	========================================================================== -->

<hr> 
<a name="changeRateChr"> 

<center>
<b> Change rate details </b><p>

<table border=1>
	<tr><th> Chromosome </th><th> Length </th><th> Changes </th><th> Change rate </th></tr> 
		<tr>
			<td> 1 </td>
			<td class="numeric"> 301,354,135 </td>
			<td class="numeric"> 23 </td>
			<td class="numeric"> 13,102,353 </td>
		</tr> 
		<tr>
			<td> OM </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 0 </td>
			<td class="numeric"> 0 </td>
		</tr> 
	<tr>
		<th> Total </th>
		<th class="numeric"> 301,354,136 </th>
		<th class="numeric"> 24 </th>
		<th class="numeric"> 12,556,422 </th>
	</tr> 
</table> 
</center>

<!--==========================================================================
	Changes by type
	========================================================================== -->

<hr> 
<a name="changesByType"> 
<center>
<b> Number changes by type</b><p>

<table border=1>
	<thead>
		<tr>
			<th> <b> Type   </b> </th>
			<th> <b> Total  </b> </th>
			<th> <b> Homo   </b> </th>
			<th> <b> Hetero </b> </th>
		</tr>
	</thead>
    <tbody>
	    <tr>
	    	<td> <b> SNP </b> </td>
	    	<td class="numeric" bgcolor="#ff0000"> 23 </td>
	    	<td class="numeric" bgcolor="#ff0000"> 10 </td>
	    	<td class="numeric" bgcolor="ffffff"> 13 </td>
	    </tr>
	    <tr>
	    	<td> <b> MNP </b> </td>
	    	<td class="numeric" bgcolor="#00ff00"> 1 </td>
	    	<td class="numeric" bgcolor="#00ff00"> 1 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> INS </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> DEL </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> MIXED </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> Interval </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
    </tbody>
    <tfoot>
	    <tr>
	    	<th><b>Total </b> </th>
	    	<th class="numeric"> 24 </th>
	    	<th class="numeric"> 11 </th>
	    	<th class="numeric"> 13 </th>
	    </tr>
    </tfoot>
</table>
</center>

<!--==========================================================================
	Effects by impact
	========================================================================== -->

<hr> 
<a name="effectsImpact"> 
<center>
<b> Number of effects by impact </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> MODERATE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 2 </td> 
			<td class="numeric" bgcolor="#00ff00"> 1.6% </td>
		</tr>
		<tr> 
			<td> <b> MODIFIER </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 123 </td> 
			<td class="numeric" bgcolor="#ff0000"> 98.4% </td>
		</tr>
</table><br>
<p>
</center>

<!--==========================================================================
	Effects by functional class
	========================================================================== -->

<hr> 
<a name="effectsFuncClass"> 
<center>
<b> Number of effects by functional class </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> MISSENSE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 2 </td> 
			<td class="numeric" bgcolor="ffffff"> 100% </td>
		</tr>
</table><br>
<p>

Missense / Silent ratio: </th><td class="numeric"> ?

</center>

<!--==========================================================================
	Effects stats
	========================================================================== -->

<hr> 
<a name="effects"> 
<center>
<b> Number of effects by type and region </b> <p> 

<table border=0>
	<tr>
		<th> Type </th>
		<th> Region </th>
	</tr>
	<tr>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> DOWNSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#22dc00"> 11 </td> 
			<td class="numeric" bgcolor="#22dc00"> 8.8% </td>
		</tr>
		<tr> 
			<td> <b> INTERGENIC </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#738b00"> 34 </td> 
			<td class="numeric" bgcolor="#738b00"> 27.2% </td>
		</tr>
		<tr> 
			<td> <b> NONE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 0.8% </td>
		</tr>
		<tr> 
			<td> <b> NON_SYNONYMOUS_CODING </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#03fb00"> 2 </td> 
			<td class="numeric" bgcolor="#03fb00"> 1.6% </td>
		</tr>
		<tr> 
			<td> <b> UPSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 74 </td> 
			<td class="numeric" bgcolor="#ff0000"> 59.2% </td>
		</tr>
		<tr> 
			<td> <b> UTR_3_PRIME </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#06f800"> 3 </td> 
			<td class="numeric" bgcolor="#06f800"> 2.4% </td>
		</tr>
</table><br>
 </td>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> DOWNSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#22dc00"> 11 </td> 
			<td class="numeric" bgcolor="#22dc00"> 8.8% </td>
		</tr>
		<tr> 
			<td> <b> EXON </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#03fb00"> 2 </td> 
			<td class="numeric" bgcolor="#03fb00"> 1.6% </td>
		</tr>
		<tr> 
			<td> <b> INTERGENIC </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#738b00"> 34 </td> 
			<td class="numeric" bgcolor="#738b00"> 27.2% </td>
		</tr>
		<tr> 
			<td> <b> NONE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 0.8% </td>
		</tr>
		<tr> 
			<td> <b> UPSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 74 </td> 
			<td class="numeric" bgcolor="#ff0000"> 59.2% </td>
		</tr>
		<tr> 
			<td> <b> UTR_3_PRIME </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#06f800"> 3 </td> 
			<td class="numeric" bgcolor="#06f800"> 2.4% </td>
		</tr>
</table><br>
 </td>
	</tr>
</table>
<p>

<img src="http://chart.apis.google.com/chart?cht=bvs&chbh=50,4,8&chs=800x300&chxt=y,y,x,x&chtt=Variations&chxp=1,50.0|3,50.0&chxr=0,0.0,59.199999999999996|1,0.0,100.0|3,0.0,100.0&chco=00FFFF,FA8072,FF0000,A52A2A,FFA500,DEB887,E9967A,7FFFD4,808080&chdl=Upstream|5%27UTR|Exon|Splice+Donor|Intron|Splice+Acceptor|3%27UTR|Downstream|Intergenic&chxl=1:|%25|2:|Intergenic|Up|5%27UTR|Exon|Donor|Intron|Acceptor|Exon|3%27UTR|Down|Intergenic|3:|&chd=e:AA__AAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAABvAAAAAABvAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAACmAAAA,AAAAAAAAAAAAAAAAAAJhAA,dZAAAAAAAAAAAAAAAAAAdZ" border=1><p>

</center>
<!--==========================================================================
	Quality plots
	========================================================================== -->

	<hr> 
	<a name="quality"> 
	<b> Quality: </b> 
	<p>
	<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>0</td> </tr>
		<tr> <th>Max</th><td>99</td> </tr>
		<tr> <th>Mean</th><td>72.083</td> </tr>
		<tr> <th>Median</th><td>99</td> </tr>
		<tr> <th>Standard deviation</th><td>33.403</td> </tr>
		<tr> <th>Values</th><td>0,38,40,99</td> </tr>
		<tr> <th>Count</th><td>1,8,1,14</td> </tr>
		</tr>
	</table>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chd=e:AAh6AA..&chg=9.0,1.0,3,2&chxp=1,50.0|3,50.0&chtt=Quality+histogram&chxl=1:|Count|2:|0|||99|3:|Quality&chxr=0,1.0,14.0|1,0.0,100.0|3,0.0,100.0"><br>
	</pre>

<!--==========================================================================
	Coverage
	========================================================================== -->

	<hr> 
	<a name="coverage"> 
	<b> Coverage:</b> 
	<p>
	<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>0</td> </tr>
		<tr> <th>Max</th><td>0</td> </tr>
		<tr> <th>Mean</th><td>0</td> </tr>
		<tr> <th>Median</th><td>0</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>0</td> </tr>
		<tr> <th>Count</th><td>24</td> </tr>
		</tr>
	</table>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Coverage+histogram&chxl=1:|Count|2:|0|3:|Coverage&chxr=0,0.0,24.0|1,0.0,100.0|3,0.0,100.0"><br>
	</pre>

<!--==========================================================================
	InDels
	========================================================================== -->


<!--==========================================================================
	Base changes
	========================================================================== -->

<hr> 
<a name="baseChages"> 
<center>
<b> Base changes (SNPs) </b> <p>

<table border=1>
    <tr>
    	<td> &nbsp; </td>  <th> <b> A </b> </th> <th> <b> C </b> </th> <th> <b> G </b> </th> <th> <b> T </b> </th>
    </tr>
	    <tr> <th> <b> A </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="#33cc00"> 2 </td><td class="numeric" bgcolor="#669900"> 3 </td><td class="numeric" bgcolor="#669900"> 3 </td> </tr>
	    <tr> <th> <b> C </b> </th><td class="numeric" bgcolor="#00ff00"> 1 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="#00ff00"> 1 </td><td class="numeric" bgcolor="#cc3200"> 5 </td> </tr>
	    <tr> <th> <b> G </b> </th><td class="numeric" bgcolor="#ff0000"> 6 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> T </b> </th><td class="numeric" bgcolor="#00ff00"> 1 </td><td class="numeric" bgcolor="#00ff00"> 1 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
</table>
</center>

<p>

<!--==========================================================================
	Ts/Tv
	========================================================================== -->

<hr> 
<a name="tstv">
<center> <b> Ts/Tv (transitions / transversions) </b> <p> </center>

<small>
<b>Note:</b> Only SNPs are used for this statistic.<br>
<b>Note:</b> This Ts/Tv ratio is a 'raw' ratio. Some people prefer to use a ratio of rates, not observed events. In that case, you need to multiply by 2.0 (since there are twice as many possible transitions than transversions, E[Ts/Tv] ratio is twice the ratio of events).
</small>
<p>

<center>
<table border=1>
	<tr> <th> Transitions </th><td class="numeric"> 15 </td> </tr>
	<tr> <th> Transversions </th><td class="numeric"> 8 </td> </tr>
	<tr> <th> Ts/Tv ratio </th><td class="numeric"> 1.875 </td> </tr>
</table>
</center>
<p>

	<b>All variants:</b>
	<pre>Sample ,Total
Transitions ,15,15
Transversions ,8,8
Ts/Tv ,1.875,1.875
</pre>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
	<pre>No results available (empty input?)</pre>
<p>

<!--==========================================================================
	Allele frequency
	========================================================================== -->

<hr> 
<a name="alleleFreq">
<center> <b> Frequency of alleles </b> <p> </center>

Note: Number of times an allele appears once (singleton), twice (doubletons), etc.<p>

	<b>All variants:</b>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Alleles+frequencies&chxl=1:|Count|2:|1|3:|Alleles&chxr=0,0.0,23.0|1,0.0,100.0|3,0.0,100.0"><br>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>1</td> </tr>
		<tr> <th>Max</th><td>1</td> </tr>
		<tr> <th>Mean</th><td>1</td> </tr>
		<tr> <th>Median</th><td>1</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>1</td> </tr>
		<tr> <th>Count</th><td>23</td> </tr>
		</tr>
	</table>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
<p>


<!--==========================================================================
	Codon change table
	========================================================================== -->
	
<hr> 
<a name="codonChanges">
<center>
<b> Codon changes</b> <p>

	<div class="note">
		How to read this table: <br>
		- Rows are reference codons and columns are changed codons. E.g. Row 'AAA' column 'TAA' indicates how many 'AAA' codons have been replaced by 'TAA' codons.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

  
<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> GAC </th>  <th> GTC </th> 
		</tr>
	</thead>
			<tr> 
				<th> GAC </th>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> GTC </th>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<th class="numeric"> &nbsp </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Amino acid change table
	========================================================================== -->

<hr> 
<a name="aaChanges">
<center>
<b> Amino acid changes</b><p> 

	<div class="note">
		How to read this table: <br>
		- Rows are reference amino acids and columns are changed amino acids. E.g. Row 'A' column 'E' indicates how many 'A' amino acids have been replaced by 'E' amino acids.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> D </th>  <th> V </th> 
		</tr>
	</thead>
			<tr> 
				<th> D </th>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> V </th>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<th class="numeric"> &nbsp </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Chromosome change table
	========================================================================== -->

<hr> 
<a name="chrChanges">
<center>
<b> Changes by chromosome</b><p> 

<center>
	<pre>
		<img src="http://chart.apis.google.com/chart?cht=lc&chs=800x300&chxt=y,y,x,x&chg=3.0E7,2.0,3,2&chxp=1,50.0|3,50.0&chtt=Changes+histogram%3A+1&chxr=0,0.0,23.0|1,0.0,100.0|3,0.0,100.0&chd=e:..AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&chxl=1:|Changes%2F10Mb|2:|0||||||||||100000000||||||||||200000000||||||||||300000000|3:|Position"><br>
		1, Position,0,10000000,20000000,30000000,40000000,50000000,60000000,70000000,80000000,90000000,100000000,110000000,120000000,130000000,140000000,150000000,160000000,170000000,180000000,190000000,200000000,210000000,220000000,230000000,240000000,250000000,260000000,270000000,280000000,290000000,300000000
1,Count,23,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0

	</pre>
	<pre>
		<img src="http://chart.apis.google.com/chart?cht=lc&chd=e:AAAA&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Changes+histogram%3A+OM&chxr=0,0.0,0.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Changes%2F1b|2:|0|1|3:|Position"><br>
		OM, Position,0,1
OM,Count,0,0

	</pre>
</center>

<!--==========================================================================
	Changes by gene 
	========================================================================== -->

<center>
	<hr>
	<b> Details by gene </b><p>
	<p>
	<b><a href="snpEff_genes.txt">Here</a></b> you can find a tab-separated table.
</center>
</div>
</center>